sys/net/gnrc/sock/udp/gnrc_sock_udp.c

```diff
@@ -24,14 +24,70 @@
 #include "net/gnrc/udp.h"
 #include "net/sock/udp.h"
 #include "net/udp.h"
-#include "random.h"
 
 #include "gnrc_sock_internal.h"
 
 #ifdef MODULE_GNRC_SOCK_CHECK_REUSE
 static sock_udp_t *_udp_socks = NULL;
 #endif
 
+static uint16_t _dyn_port_next = 0;
+
+/* --------------------------------- helper --------------------------------- */
+
+/**
+ * @brief   Checks if a given UDP port is already used by another sock
+ */
+static bool _dyn_port_used(uint16_t port)
+{
+#ifdef MODULE_GNRC_SOCK_CHECK_REUSE
+    for (sock_udp_t *ptr = _udp_socks; ptr != NULL;
+         ptr = (sock_udp_t *)ptr->reg.next) {
+        bool spec_addr = false;
+        for (unsigned i = 0; i < sizeof(ptr->local.addr); i++) {
+            const uint8_t *const p = (uint8_t *)&ptr->local.addr;
+            if (p[i] != 0) {
+                spec_addr = true;
+            }
+        }
+        if (spec_addr) {
+            continue;
+        }
+        if (ptr->local.port == port) {
+            /* port already in use by another sock */
+            return true;
+        }
+    }
+#else
+    (void) port;
+#endif /* MODULE_GNRC_SOCK_CHECK_REUSE */
+    return false;
+}
+
+/**
+ * @brief   returns a UDP port, and checks for reuse if required
+ *
+ * complies to RFC 6056, see https://tools.ietf.org/html/rfc6056#section-3.3.3
+ */
+static uint16_t _get_dyn_port(sock_udp_t *sock)
+{
+    uint16_t port;
+    uint16_t count = GNRC_SOCK_DYN_PORTRANGE_NUM;
+    do {
+        port = GNRC_SOCK_DYN_PORTRANGE_MIN +
+               (_dyn_port_next + GNRC_SOCK_DYN_PORTRANGE_OFF) % GNRC_SOCK_DYN_PORTRANGE_NUM;
+        _dyn_port_next++;
+        if ((sock == NULL) || (sock->flags & SOCK_FLAGS_REUSE_EP) ||
+                              !_dyn_port_used(port)) {
```

Comment:
@OlegHahm it is, because that's an **or** expression, or what do you suggest/mean?